### PR TITLE
[ONNX] Capture annotated attributes for local function

### PR DESCRIPTION
**Pull request by BowenBao**
Enables local function export to capture annotated attributes.
For example:
```python
class M(torch.nn.Module):
    num_layers: int

    def __init__(self, num_layers):
        super().__init__()
        self.num_layers = num_layers

    def forward(self, args):
        ...
```
`num_layers` will now be captured as attribute of local function `M`.

### CODE REVIEW

**Comment by BowenBao** [comment]
cc @garymm please take a look.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 1]
```diff
@@ -1020,6 +1071,28 @@ NodeAttrNameMap FunctionExtractor::run() {
...
+  // is popped from TracingState.
+  // This function inspects the node, and its subblock, to find
+  // the node associated with the most recent ScopePtr.
+
+  // Skip the "real" last node which is `return_node`.
+  TORCH_INTERNAL_ASSERT(forward_node->kind() == prim::TracedModuleForward);
+  auto* block = forward_node->blocks()[0];
+  for (auto* node : block->nodes().reverse()) {
+    if (node->kind() == prim::TracedModuleForward) {
+      auto* target_node = GetNodeOfMostRecentScope(node);
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/function_extraction.cpp:1087]: In general, if the type name is pretty short and simple, I think it's a bit more readable to use it rather than `auto`.
In this case I think `Node*` is preferrable to `auto*`.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 2]
```diff
@@ -1020,6 +1071,28 @@ NodeAttrNameMap FunctionExtractor::run() {
...
+Node* GetNodeOfMostRecentScope(Node* forward_node) {
+  // This function is used to retrieve the node representing the most recent
+  // ScopePtr. This function should only be invoked from module forward hook. At
+  // this point, module forward call is completed, and the most recent ScopePtr
+  // is popped from TracingState.
+  // This function inspects the node, and its subblock, to find
+  // the node associated with the most recent ScopePtr.
+
+  // Skip the "real" last node which is `return_node`.
+  TORCH_INTERNAL_ASSERT(forward_node->kind() == prim::TracedModuleForward);
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/function_extraction.cpp:1083]: Can we log something about `forward_node` here so the error message is useful in case this fails?

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 3]
```diff
@@ -1045,6 +1118,59 @@ NodeAttrNameMap ONNXFunctionExtraction(
   return fe.run();
 }
 
+Node* ONNXGetPreviousScope(std::shared_ptr<Graph>& graph) {
+  auto* last_node = graph->nodes().back()->prev();
+  auto* scope_node = GetNodeOfMostRecentScope(last_node);
+  auto* attr_node = scope_attr_graph_->create(prim::TracedModuleForward);
+  attr_node->setScope(scope_node->scope());
+  TORCH_INTERNAL_ASSERT(
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/function_extraction.cpp:1126]: Can we log something about `scope_node` here so the error message is useful in case this fails?

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -665,10 +669,52 @@ def _find_typename(v):
...
 def _reset_trace_module_map():
     torch.jit._trace._trace_module_map = None
+    torch._C._jit_pass_onnx_clear_scope_records()
+
+def _get_module_attributes(module):
+    from typing import get_type_hints
+    annotations = get_type_hints(type(module))
+    base_m_annotations = get_type_hints(torch.nn.Module)
+    if annotations and base_m_annotations:
+        [annotations.pop(k, None) for k in base_m_annotations.keys()]
```
Comment (garymm) [inline torch/onnx/utils.py:687]: I believe by default iterating over a dict yields its keys so you can comit `.keys()`:

```suggestion
        [annotations.pop(k, None) for k in base_m_annotations]
```

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -665,10 +669,52 @@ def _find_typename(v):
...
+    else:
+        annotations = {}
+    return {k : getattr(module, k) for k in annotations.keys()}
+
+def _register_module_forward_hook(module, hook):
+    for child in module.children():
```
Comment (garymm) [inline torch/onnx/utils.py:693]: Can we use `module.modules()` to handle the recursion for us?
https://discuss.pytorch.org/t/module-children-vs-module-modules/4551

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -665,10 +669,52 @@ def _find_typename(v):
...
+def _track_module_attributes_forward_pre_hook(module, input):
+    module._onnx_attrs = _get_module_attributes(module)
+
+def _track_module_attributes_forward_hook(module, input, output):
+    tracing_state = torch._C._get_tracing_state()
+    if not tracing_state:
+        return
+
+    graph = tracing_state.graph()
+    attr_name = "_onnx_attrs"
```
Comment (garymm) [inline torch/onnx/utils.py:711]: Let's share this string between here and `_track_module_attributes_forward_pre_hook` to make the relationship clear.

**Review by garymm — CHANGES_REQUESTED** [review]
As you can see from my questions I don't really understand some of this, so it may be faster to schedule a pair programming session. Feel free to do so any time on Friday (even though my calendar says I'm busy).

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 7]
```diff
@@ -1020,6 +1071,28 @@ NodeAttrNameMap FunctionExtractor::run() {
...
 
+Node* GetNodeOfMostRecentScope(Node* forward_node) {
+  // This function is used to retrieve the node representing the most recent
+  // ScopePtr. This function should only be invoked from module forward hook. At
+  // this point, module forward call is completed, and the most recent ScopePtr
+  // is popped from TracingState.
+  // This function inspects the node, and its subblock, to find
+  // the node associated with the most recent ScopePtr.
+
+  // Skip the "real" last node which is `return_node`.
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/function_extraction.cpp:1082]: good catch forgot to remove.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/function_extraction.cpp:1087]: I agree, especially for some ambiguous function returns that is not obvious about return type. For this particular case though, both function and variable name should reflect the type is `Node`, I feel it should be ok with `auto`. Let me know if you still prefer `Node*` here.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 8]
```diff
@@ -9,6 +9,12 @@ namespace {
 
 using scope_list = std::vector<ScopePtr>;
 
+// Predefined attributes retrieved from module.
+// They are not used by subgraph nodes, so it only works as an annotation to
+// help with later kernel replacement. It does not affect subgraph execution.
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/function_extraction.cpp:14]: thanks, this is better

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 9]
```diff
@@ -487,14 +492,50 @@ Node* FunctionExtractor::CreateFunctionDefNode(
...
 
+  // Set predefined attributes
+  std::unordered_set<Symbol> predefined_attr_names;
+  bool first_iteration = true;
+  for (const auto& it : func_ctx.scope_ctxs_) {
+    auto scope = it.first;
+    auto predefined_attr_node = scope_attr_map_.find(scope);
+    if (predefined_attr_node != scope_attr_map_.end()) {
+      auto names = predefined_attr_node->second->attributeNames();
+      if (first_iteration) {
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/function_extraction.cpp:511]: It could be empty after the first iteration.

File: test/onnx/test_utility_funs.py [hunk 10]
```diff
@@ -802,6 +802,46 @@ def forward(self, x):
         funcs = onnx_model.functions
         self.assertIn("M", [f.name for f in funcs])
 
+    @skipIfUnsupportedMinOpsetVersion(15)
+    def test_local_function_predefined_attributes(self):
+        class M(torch.nn.Module):
+            num_layers: int
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_utility_funs.py:808]: These are very good questions. I found the explanation/definition for class annotations here https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations. 
> Does this need to be defined here rather than just in __init__? If so, why?

Yes, because [get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) tries to `Return a dictionary containing type hints for a function, method, module or class object`, not module/class instance.
And in pep-526, setting annotation inside `__init__` alone sets annotation for instance attribute, instead of class attribute.

> Does it have to do with type hints?

Yes, this is explained in pep-526.

> Does setting a type hint in __init__ not work?

No, same as first question. 

> If type hints are required, why are they?
> I would imagine the tracer would assign a type to the attribute even if there's no type hint?

Tracer does not pick up these attributes. We pick up these attributes by inspecting type of module in forward pre hook. So it's purely python and unrelated to tracer nor PyTorch. 

> Could you document this in the doc string for export_modules_as_functions in torch.onnx.export?

Adding.

**Comment by BowenBao** [comment]
@garymm I'm working on adding comments and updating code according to comments. I'll ping again once I have finished, and let me know again if it is clearer or a meeting is needed for clarification.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 11]
```diff
@@ -487,14 +492,50 @@ Node* FunctionExtractor::CreateFunctionDefNode(
     auto* n = n_it.first;
     for (const auto& attr_it : n_it.second) {
       const auto& attr = attr_it.first;
-      auto attr_name =
-          std::string(n->kind().toUnqualString()) + '_' + attr.toUnqualString();
+      auto attr_name = "inferred::" + std::string(n->kind().toUnqualString()) +
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/function_extraction.cpp:495]: Adding  

      // Add prefix "inferred::" to name of inferred attribute.
      // This is to differentiate from predefined attributes picked up
      // from python module annotation.

File: torch/onnx/utils.py (hunk 6 above)
Comment (BowenBao) [reply torch/onnx/utils.py:711]: moved things around a bit, making the hooks local functions.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/function_extraction.cpp:500]: changed "predefined" -> "annotated"

**Comment by BowenBao** [comment]
@garymm updated PR. Added doc in __init__.py, also renamed all `predefined attributes` to `annotated attributes`.

File: torch/csrc/jit/passes/onnx/function_extraction.cpp [hunk 12]
```diff
@@ -9,6 +9,14 @@ namespace {
 
 using scope_list = std::vector<ScopePtr>;
 
+// Annotated attributes retrieved from module by inspecting module annotations.
+// These attributes are not used inside the subgraph of ONNX local function
+// because they are not created by PyTorch JIT tracking, but they may be used by
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/function_extraction.cpp:14]:
```suggestion
// because they are not created by PyTorch JIT tracing, but they may be used by
```

File: torch/onnx/__init__.py [hunk 13]
```diff
@@ -304,6 +304,17 @@ def forward(self, x):
             particular types of modules to export as local functions in ONNX.
             This feature requires ``opset_version`` >= 15, otherwise the export will fail. This is because
             ``opset_version`` < 15 implies IR version < 8, which means no local function support.
+            Module variables will be exported as function attributes. There are two categories of function
+            attributes.
+
+            1. Annotated attributes: class variables that are annotated will be exported as attributes.
```
Comment (garymm) [inline torch/onnx/__init__.py:310]:
```suggestion
            1. Annotated attributes: class variables that have type annotations be exported as attributes.
```

File: torch/onnx/__init__.py [hunk 14]
```diff
@@ -304,6 +304,17 @@ def forward(self, x):
...
+            Annotated attributes are not used inside the subgraph of ONNX local function because
+            they are not created by PyTorch JIT tracking, but they may be used by consumers
```
Comment (garymm) [inline torch/onnx/__init__.py:312]:
```suggestion
            they are not created by PyTorch JIT tracing, but they may be used by consumers
```

**Review by garymm — APPROVED** [review]
Looks good!
I added the `release notes` and `topic` labels. I think if we do this consistently going forward it will be much easier to produce release notes, so please try to do that when approving PRs.

File: torch/onnx/__init__.py (hunk 13 above)
Comment (BowenBao) [reply torch/onnx/__init__.py:310]: thx, I'm adding the link of PEP 526 to provide extra clarity.

            1. Annotated attributes: class variables that have type annotations via
            `PEP 526-style <https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations>`_
            will be exported as attributes.
